mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h

```diff
@@ -516,6 +516,28 @@ LogicalResult vectorizeOpPrecondition(Operation *op,
 
 using LinalgLoops = SmallVector<Operation *, 4>;
 
+// Forward declaration
+struct ControlDropUnitDims;
+
+/// Collapse the given \p value to \p targetShape. The \p reassociation is used
+/// when `rankReductionStrategy` of \p control is set to
+/// `RankReductionStrategy::ReassociativeReshape`. Will return failure if the
+/// operand has memref type with a non-identity layout or tensor type with an
+/// encoding.
+FailureOr<Value> collapseValue(RewriterBase &rewriter, Location loc,
```

Comment:
Hmm, on second thought, I think it might be better to make this method a static value in `ControlDropUnitDims` since this only is relevant for drop unit-dims. So something like

```
struct ControlDropUnitDims {

private:
  static CollapseFnTy defaultCollapseValueImpl;

  static ExpandFnTy defaultExpandValueImpl;
};
```

Reply:
I've turned them into private `static` class functions. Using a `static` `std::function` member would require a global constructor, triggering a warning.